jupyter_scheduler/orm.py

```diff
@@ -78,37 +93,52 @@ def process_result_value(self, value, dialect):
 mapper_registry = registry()
 
 
-class Job(Base):
-    __tablename__ = "jobs"
-    job_id = Column(String(36), primary_key=True, default=generate_uuid)
+@declarative_mixin
+class CommonColumns:
     runtime_environment_name = Column(String(256), nullable=False)
     runtime_environment_parameters = Column(JsonType(1024))
     compute_type = Column(String(256), nullable=True)
     input_uri = Column(String(256), nullable=False)
     output_prefix = Column(String(256))
     output_formats = Column(JsonType(512))
-    output_uri = Column(String(256), default=output_uri)
     name = Column(String(256))
-    job_definition_id = Column(String(36))
     idempotency_token = Column(String(256))
-    status = Column(String(64), default=Status.STOPPED)
     tags = Column(JsonType(1024))
-    status_message = Column(String(1024))
-    start_time = Column(Integer)
-    end_time = Column(Integer)
     parameters = Column(JsonType(1024))
-    url = Column(String(256), default=generate_url)
     email_notifications = Column(EmailNotificationType(1024))
     timeout_seconds = Column(Integer, default=600)
     retry_on_timeout = Column(Boolean, default=False)
     max_retries = Column(Integer, default=0)
     min_retry_interval_millis = Column(Integer, default=0)
     output_filename_template = Column(String(256))
-    pid = Column(Integer)
     update_time = Column(Integer, default=get_utc_timestamp, onupdate=get_utc_timestamp)
     create_time = Column(Integer, default=get_utc_timestamp)
 
 
+class Job(CommonColumns, Base):
+    __tablename__ = "jobs"
+    job_id = Column(String(36), primary_key=True, default=generate_uuid)
+    job_definition_id = Column(String(36))
+    output_uri = Column(String(256), default=output_uri)
+    status = Column(String(64), default=Status.STOPPED)
+    status_message = Column(String(1024))
+    start_time = Column(Integer)
+    end_time = Column(Integer)
+    url = Column(String(256), default=generate_jobs_url)
+    pid = Column(Integer)
+
+
+class JobDefinition(CommonColumns, Base):
+    __tablename__ = "job_definitions"
+    job_definition_id = Column(String(36), primary_key=True, default=generate_uuid)
+    schedule = Column(String(256))
+    timezone = Column(String(36))
+    url = Column(String(256), default=generate_job_definitions_url)
+    create_time = Column(Integer, default=get_utc_timestamp)
+    next_run_time = Column(Integer, default=next_run_time)
```

Comment:
I don't think the next run time should be included in the DB record, since this value is derived from the schedule and timezone, which means we're having duplicate state. To avoid issues with keeping the state self-consistent, why don't we just compute it dynamically whenever we need it? This also has performance issues; if the transactions are immediately committed, we're writing to disk every time a job definition executes a job (since we need to update the next run time).

Reply:
My next step to this PR is adding a task runner that is going to query the job definition table every 30s and query for definitions that have `next_run_time > now - 30s`. The task runner will then create jobs for job definitions returned in this query; this is a rather simple mechanism to fetch the next pool of tasks. 

Computing which jobs to create dynamically would require querying every job definition, and then iterating and computing the `next_run_time` for each of them. 

I see your point that the current mechanism will write to disk (and same table) when a job with definition completes, but that is happening already today because we update status, end_time, failure etc during a job lifecycle; in addition these updates are localized to specific rows. With realistic no (1k - 10k) of jobs, I doubt that we are going to hit these performance limits.

Reply:
I am not too concerned with the duplicate state in this case and I do think it will be helpful to helpful to have a simple way of querying for jobs that need to be run in the current iteration of the event loop.